Add Portuguese language support to client

davidklaw opened the pull request:
This is a continuation of #656, which appears to be a standstill.

I fixed the compilation errors and added or revised some of the translations if missing or ill-fitting. I adjusted layouts where shorter translations weren't possible or felt undesirable.

For visual comparison, I compiled these sheets of English vs. Portuguese:

![image](https://github.com/user-attachments/assets/63cc7f63-4fc6-4016-beb9-ad1f410d98b7)
![image](https://github.com/user-attachments/assets/cc5b8fea-4c14-4c89-8a3b-4e98bb5e87bb)
![image](https://github.com/user-attachments/assets/4fe4b027-5758-4809-b1db-8415a29caaae)
![image](https://github.com/user-attachments/assets/340459e9-cfde-41dd-bc3a-caa2278f7908)
![image](https://github.com/user-attachments/assets/f477617e-b457-4f50-b741-b2e866beccf0)
![image](https://github.com/user-attachments/assets/3b826f6f-90e7-4f26-b331-5c2c7bd0d7d2)

This closes #656

akirmse commented on module/mailnews/mailfile.c:
```diff
@@ -310,8 +310,7 @@ Bool MailParseMessage(int msgnum, MailInfo *info)
    char filename[MAX_PATH + FILENAME_MAX];
    char line[MAX_LINE];
    int num_fields = 5;  // Don't increase this without changing szLoadString (only 5 at a time)
```
This looks very suspicious, as there are no longer 5 fields.

akirmse commented on module/mailnews/mailfile.c:
```diff
@@ -405,8 +405,7 @@ Bool MailParseMessageHeader(int msgnum, char *filename, MailHeader *header)
 {
    FILE *infile;
    int num_fields = 5;  // Don't increase this without changing szLoadString (only 5 at a time)
```
Here too.

akirmse reviewed (commented):
Please fix the PR title.

akirmse commented on module/mailnews/mailfile.c:
```diff
@@ -161,9 +161,13 @@ void MailNewMessage(int server_index, char *sender, int num_recipients,
+            subject_found = False;
+            subject = "";
+         }
       }
```
Now that there are 3 of these, let's make this into a loop over the ids instead.  Let me know if you want help with that.

akirmse reviewed (commented):
Please change the title of this PR to something meaningful.

davidklaw commented:
In addition to making changes on feedback, I also discovered I hadn't review and addressed issues with the chess and character creation dialogs. Those have been now been reviewed and updated.

akirmse commented on module/mailnews/mailfile.c:
```diff
@@ -428,12 +439,12 @@ Bool MailParseMessageHeader(int msgnum, char *filename, MailHeader *header)
             line[strlen(line) - 1] = 0;
             
-            ptr = line + strlen(fields[i]);
+            ptr = line + strlen(fields[i].c_str());
```
strlen(fields[i].c_str()) -> fields[i].size()

akirmse commented on module/mailnews/mailfile.c:
```diff
@@ -309,10 +319,10 @@ Bool MailParseMessage(int msgnum, MailInfo *info)
-   int field_ids[] = {IDS_SUBJECT_ENGLISH, IDS_SUBJECT_GERMAN,
-                      IDS_FROM, IDS_TO, IDS_DATE};
-   char *fields[5], *ptr;
+   int field_ids[] = {IDS_SUBJECT_ENGLISH, IDS_SUBJECT_GERMAN, IDS_SUBJECT_PORTUGUESE, IDS_FROM, IDS_TO, IDS_DATE};
```
field_ids[] should be const

akirmse commented on module/mailnews/mailfile.c:
```diff
@@ -126,12 +128,15 @@ void MailNewMessage(int server_index, char *sender, int num_recipients,
    int index, num_msgs, msgnum;
    char new_msg[MAXMAIL + 200 + MAX_SUBJECT + MAXUSERNAME * MAX_RECIPIENTS];
-   char *subject, *ptr, *subject_str;
+   int subject_ids[] = {IDS_SUBJECT_ENGLISH, IDS_SUBJECT_GERMAN, IDS_SUBJECT_PORTUGUESE};
```
Make const

davidklaw replied:
The likelihood of 3 more language anytime soon seems low, but would you like me to add support for more than 5 strings here?

akirmse replied:
Yeah, we should avoid relying on those terrible static bufferes.

davidklaw replied:
I gave it a go.

davidklaw commented:
Pulling it back to draft. Replies are populating the subject field with the date :/

akirmse commented on clientd3d/client.rc:
```diff
@@ -18,6 +18,754 @@
+            POPUP "Gr�ficos de barra"
+            BEGIN
+                MENUITEM "&Barras normais",              3500
+                MENUITEM "&Barras completas",             3501
```
We can't have the same hotkey (B, in this case) for multiple items in the same menu.  I'm afraid this is going to require a thorough audit.

davidklaw replied:
I've gone through and updated hotkeys by comparing available hotkeys to the English version and ensuring that no duplicate hotkeys exist within the same level of each menu.

davidklaw commented:
I noticed that Portguguese and German translations more often than not, use the numerical equivalents of constant ids (e.g. `-1` vs. `IDC_STATIC`, `IDCANCEL`, etc) . Should the Portuguese resources use these as well?

Here's what I mean
```
BEGIN
    GROUPBOX        "Confirmação",-1,7,28,228,33
    LTEXT           "Confirmar esta ação irá destruir o seu personagem e criar um novo. É de sua responsabilidade iniciar um novo jogo agora.",-1,7,4,235,18
    LTEXT           "Para confirmação esta ação, insira sua senha:",-1,14,40,106,19
    EDITTEXT        1260,131,42,97,12,ES_PASSWORD | ES_AUTOHSCROLL
    DEFPUSHBUTTON   "Confirmar",1,7,65,107,14
    PUSHBUTTON      "Cancelar",2,120,65,115,14
END
```

```
BEGIN
    GROUPBOX        "Verification",IDC_STATIC,7,38,228,43
    LTEXT           "Performing a reroll will destroy your character, and will create a new character for you to start over.  You are responsible for your own choice to begin the game again.",IDC_STATIC,7,7,228,27
    LTEXT           "To verify that you wish to do this, enter your current account password here.",IDC_STATIC,14,50,114,26
    EDITTEXT        IDC_SUICIDE_PASSWORD,131,52,97,12,ES_PASSWORD | ES_AUTOHSCROLL
    DEFPUSHBUTTON   "Reroll",IDOK,131,74,46,14
    PUSHBUTTON      "Cancel",IDCANCEL,184,74,46,14
END
```

Meridian59 commented:
No, the names are better.  If the #defines for these ever change, all those
translations will break.  I don't know why it was done this way.

On Sat, Sep 21, 2024 at 4:18 AM David Klawitter ***@***.***>
wrote:

> I noticed that Portguguese and German translations more often than not,
> use the numerical equivalents of constant ids (e.g. -1 vs. IDC_STATIC,
> IDCANCEL, etc) . Should the Portuguese resources use these as well?
>
> Here's what I mean
>
> BEGIN
>     GROUPBOX        "Confirmação",-1,7,28,228,33
>     LTEXT           "Confirmar esta ação irá destruir o seu personagem e criar um novo. É de sua responsabilidade iniciar um novo jogo agora.",-1,7,4,235,18
>     LTEXT           "Para confirmação esta ação, insira sua senha:",-1,14,40,106,19
>     EDITTEXT        1260,131,42,97,12,ES_PASSWORD | ES_AUTOHSCROLL
>     DEFPUSHBUTTON   "Confirmar",1,7,65,107,14
>     PUSHBUTTON      "Cancelar",2,120,65,115,14
> END
>
> BEGIN
>     GROUPBOX        "Verification",IDC_STATIC,7,38,228,43
>     LTEXT           "Performing a reroll will destroy your character, and will create a new character for you to start over.  You are responsible for your own choice to begin the game again.",IDC_STATIC,7,7,228,27
>     LTEXT           "To verify that you wish to do this, enter your current account password here.",IDC_STATIC,14,50,114,26
>     EDITTEXT        IDC_SUICIDE_PASSWORD,131,52,97,12,ES_PASSWORD | ES_AUTOHSCROLL
>     DEFPUSHBUTTON   "Reroll",IDOK,131,74,46,14
>     PUSHBUTTON      "Cancel",IDCANCEL,184,74,46,14
> END
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Meridian59/Meridian59/pull/955#issuecomment-2365149985>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AARJPIJDIFULXTZQN26QTF3ZXVIXTAVCNFSM6AAAAABOMDH332VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDGNRVGE2DSOJYGU>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

davidklaw commented:
Back to draft as I comb through and updated the use of names.

davidklaw commented:
This is ready for review. I briefly moved it to draft and back just as a safety measure, but it was unnecessary.

The latest commits address a few misses:
- Fixed the use of constant numerals rather than their names (for Portuguese only)
- Made Portuguese mail and and news lists sortable by removing the nosort flag
- Made small refinements to poor label dimensions and positions in the character creation flow
